src/plugins/controls/public/hooks/use_field_formatter.ts

```diff
@@ -29,10 +31,13 @@ export const useFieldFormatter = ({
       if (!dataViewId || !fieldSpec) return;
       // dataViews are cached, and should always be available without having to hit ES.
       const dataView = await getDataViewById(dataViewId);
-      setFieldFormatter(
-        () =>
-          dataView?.getFormatterForField(fieldSpec).getConverterFor('text') ??
-          ((toFormat: string) => toFormat)
+      const formatterForField =
+        dataView?.getFormatterForField(fieldSpec).getConverterFor('text') ??
+        ((toFormat: string) => toFormat);
+      setFieldFormatter(() =>
+        fieldSpec.type === 'date' // date fields must be passed in as a number for the formatter to work
```

Comment:
Instead of doing the string conversion on the server and then having to have custom handling for edge cases in the client, what about keeping the server code as-is and then converting formatted values to string on the client?

Reply:
I thought about this, but it felt messier to have the server return an unknown type? Right now, our server-side code is typed so that it is **supposed** to return only strings, but we never did any casting to ensure this - so it was returning numbers for both numeric and date fields despite what the typing suggested. To keep this type safe, it felt safer to me to do the casting on the server side because of this - otherwise, I would have to go around and change the typing to "unknown" everywhere 🤷 What do you think?

Reply:
Since the data may be needed in its original form, we should not cast until the last minute. I would type the `value` as `string | number` and then cast to string before passing the value or formatted value to options list.

@ThomThomson What do you think?

Reply:
That's fair - I actually originally had `string | number` but I bounced to this idea instead because it felt cleaner to have the casting at a single point (server side) rather than all over the client-side code. I'm down to make that change if @ThomThomson agrees, though!

Reply:
I'm not a huge fan of this extra layer of cast before the formatter. I'd recommend having the serverside code return `string | number` - which will likely cause a pretty large number of type failures clientside - then have those resolved by some `typeof` check before they end up in the formatter.

IMO the choice is between...

**A:** ES `string | number` -> server `string` -> client uses string or casts to number for dates
**B:** ES `string | number` -> server `string | number` -> client narrows to string or number

B seems more straightforward to me because it maintains the real types further along... 

> all over the client-side code

What other places do you anticipate needing this type narrowing? Maybe this is something that only the Options List Embeddable code needs to handle when the server returns results?

Reply:
Resolved - see https://github.com/elastic/kibana/pull/188789#issuecomment-2246338726 :+1: